ui/components/app/detected-token/detected-token.js

```diff
@@ -0,0 +1,109 @@
+import React, { useState } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import { useHistory } from 'react-router-dom';
+
+import {
+  importTokens,
+  ignoreTokens,
+  setNewTokensImported,
+} from '../../../store/actions';
+import { getDetectedTokensInCurrentNetwork } from '../../../selectors';
+
+import { isEqualCaseInsensitive } from '../../../../shared/modules/string-utils';
+import { DEFAULT_ROUTE } from '../../../helpers/constants/routes';
+import DetectedTokenSelectionPopover from './detected-token-selection-popover/detected-token-selection-popover';
+import DetectedTokenIgnoredPopover from './detected-token-ignored-popover/detected-token-ignored-popover';
+
+const DetectedToken = () => {
+  const dispatch = useDispatch();
+  const history = useHistory();
+
+  const detectedTokens = useSelector(getDetectedTokensInCurrentNetwork);
+
+  const [selectedTokens, setSelectedTokens] = useState(detectedTokens);
```

Comment:
[Suggestion] You might consider changing the shape of this state object from an array to an object, keyed by address. If you do this then you can avoid array lookups everywhere down the chain where you're dealing with a single token. You can also avoid, in most cases, passing this state array down the tree and instead derive the specific values you need from it and pass those. Your state initialization would look like

```js
const [selectedTokens, setSelectedTokens] = useState(() => _.chain(detectedTokens).keyBy('address').mapValues(() => true).value())
```

chain is a lodash method that creates a lodash functional chain around a value. keyBy creates an object by iterating over the value and using the suppplied key to use as the key for the new object, mapValues iterates over an object and updates the values, and .value() is the end of the chain returning the value computed thus far.

Another way to write this without lodash:

```js
const [selectedTokens, setSelectedTokens] = useState(() => detectedTokens.reduce((tokenObj, token) => {
  tokenObj[token.address] = true;
  return tokenObj;
}, {}));
```

Either way is acceptable. 

Now, later in your code you'll have to do things like where you were previously doing ```dispatch(importTokens(selectedTokens))``` you'll do ```dispatch(importToken(Object.keys(selectedTokens).filter(address => selectedTokens[address]).map(address => detectedTokens[address]);```


but also this:

```js
  const handleTokenSelection = (token) => {
    let newSelectedTokens = [...selectedTokens];
    let newUnSelectedTokens = [...unSelectedTokens];

    if (
      newSelectedTokens.find(({ address }) =>
        isEqualCaseInsensitive(address, token.address),
      )
    ) {
      newSelectedTokens = newSelectedTokens.filter(
        ({ address }) => address !== token.address,
      );
      newUnSelectedTokens.push(token);
    } else {
      newSelectedTokens.push(token);
      newUnSelectedTokens = newUnSelectedTokens.filter(
        ({ address }) => address !== token.address,
      );
    }
    setSelectedTokens(newSelectedTokens);
    setUnSelectedTokens(newUnSelectedTokens);
  };
```

can become

```js
  const handleTokenSelection = (token) => {
    setSelectedTokens((prevSelectedTokens => ({
      ....prevSelectedTokens,
     [token.address]: prevSelectedTokens[token.address] ? false : true,
   )}
 };
```

You can then get rid of the 'unselectedTokens' state because that is just the token addresses that are currently false.

Let me know if that makes sense? Also in things like detected-token-values you don't have to pass in the selectedTokens state and do a lookup. You can simply check if the address of the token is 'true' of 'false' in the selected token state and pass in the value of whether its checked or not, greatly simplifying logic.